platform-operator/controllers/verrazzano/component/rancher/rancher.go

```
@@ -269,161 +269,112 @@ func isRancherReady(ctx spi.ComponentContext) bool {
-		if err != nil {
-			return err
-		}
-		defer logStream.Close()
-
-		// Search the stream for the expected text
-		restartPod := false
-		scanner := bufio.NewScanner(logStream)
-		for scanner.Scan() {
-			token := scanner.Text()
-			if strings.Contains(token, "[ERROR] available chart version") {
-				log.Infof("Rancher IsReady: Failed to find system chart for pod %s: %s", pod.Name, token)
-				restartPod = true
-				break
-			}
-		}
-
-		// If the pod is failing to find the system chart for rancher-webhook, the wrong helm charts are
-		// being used by the Rancher pod. Restart the pod. This will cause another Rancher pod to become the leader,
-		// and, if needed, will recreate the custom resources related to the helm charts.
-		if restartPod {
-			// Delete custom resources containing helm charts to use
-			err := deleteClusterRepos(log)
-			if err != nil {
-				return err
-			}
-			log.Infof("Rancher IsReady: Restarting pod %s", pod.Name)
-			err = c.Delete(ctx, &podList.Items[i])
-			if err != nil {
-				return err
-			}
-			podsRestarted = true
-		}
+	log.Infof("Scaling down Rancher deployment %v", namespacedName)
+	zero := int32(0)
+	deployment.Spec.Replicas = &zero
+	if err := c.Update(context.TODO(), &deployment); err != nil {
+		return log.ErrorfNewErr("Failed to scale Rancher deployment %v to zero replicas: %v", namespacedName, err)
 	}
 
-	// If any pods were restarted, return an error so that the IsReady check will not continue
-	// any further.  Checks will resume again after the pod is ready again.
-	if podsRestarted {
-		return fmt.Errorf("Rancher IsReady: pods were restarted, waiting for them to be ready again")
+	// get the deployment and check the status to see if all replicas are down, if not we'll reconcile again
+	// until all replicas are down
+	if err := c.Get(context.TODO(), namespacedName, &deployment); err != nil {
+		return client.IgnoreNotFound(err)
```

Could this be accomplished by inducing another reconcile? In your testing, has this ever succeeded or have we had to reconcile regardless. Maybe we could return a retryable error as the default and only return nil when the first `deployment.Status.AvailableReplicas == 0` returns true (as you have in the beginning)

---

Yes I think that would work. I'd also have to add a check to only update the deployment if the replicas is not already zero, but what you're suggesting should work.

---

OK made that update.